keps/2941-DRA-Structured-Parameters/README.md

```diff
@@ -0,0 +1,416 @@
+   to a single claim will be used against quota.
+
+
+## Design Details
+
+### Resource Quota API
+
+```golang
+type ResourceQuota struct {
+    // kind is used to configure if this is for a DRA Device. Its value will be DeviceClass.
+    // +featureGate=KueueDynamicResourceAllocation
+    Kind *string `json:"kind,omitempty"`
+    
+    // deviceClassNames lists the names of all the device classes that will count against
+    // the quota defined in this resource quota object.
+    // +listType=atomic
+    // +optional
+    // +featureGate=KueueDynamicResourceAllocation
+    DeviceClassNames []corev1.ResourceName `json:"deviceClassNames,omitempty"`
+}
+```
+
+Kind field in ResourceQuota allows Kueue to distinguish between a Core resource and a DeviceClass. When the value of kind
+field is configured to be DeviceClass, the `name` becomes a canonical name of the collection of resources that can be 
+provisioned for each device class present in `deviceClassNames` field.
+
+
+With this, a ClusterQueue could be defined as follows:
+
+```yaml
+apiVersion: kueue.x-k8s.io/v1beta1
+kind: ClusterQueue
+metadata:
+  name: "cluster-queue"
+spec:
+  namespaceSelector: {} # match all.
+  resourceGroups:
+  - coveredResources: ["cpu", "memory", "gpu.example.com"]
+    flavors:
+    - name: "default-flavor"
+      resources:
+      - name: "cpu"
+        nominalQuota: 9
+      - name: "memory"
+        nominalQuota: "200Mi"
+      - name: "single-gpu"
+        deviceClassNames: ["gpu.example.com"]
+        nominalQuota: 2
+        kind: "DeviceClass"
+      - name: "shared-gpus"
+        deviceClassNames: 
+        - "ts-shared.gpu.example.com"
+        - "sp-shared.gpu.example.com"
+        nominalQuota: 2
+        kind: "DeviceClass"
+```
+
+The above ClusterQueue is an example configuration of a queue, with half quota configured for single allocation of example
+GPUs, and half quota configured for GPUs that are shared by workloads. Similarly, when DRAPartitionableDevices feature
+is supported in kubernetes, GPUs partitions can be represented by a single device class. This allows for quota to be 
+configured in Kueue without any modifications. 
+
+### Workloads
+
+When a user submits a workload and KueueDynamicResourceAllocation feature gate is on, Kueue will do the following:
+
+1. Claims will be read from resources.claims in the PodTemplateSpec.
+2. ResourceClaimSpec will be looked up either by using:
+     1. the name of the ResourceClaimTemplate or
+     2. the name of the ResourceClaim
+   
+   Both ResourceClaimTemplate or ResourceClaim will be in the same namespace as the workload.
+3. From the ResourceClaimSpec, the deviceClassName will be read.
+4. Every claim, deviceClassName for each request will be looked at 
+   1. For the workload a deviceClassMap will be created, which is map of deviceClass -> canonical name in cluster queue
+   2. for each device class the canonical quota name will be looked up and resource will be counted against it.
```

Comment:
Is it possible and desirable to add an extra check here for ResourceClaims that the claim is not already allocated?

e.g. If one workload references a claim, then it should count against the quota since it's asking for new devices. If another workload references the same claim which is now allocated, I would expect that claim not to count against the quota since no additional devices are being allocated.

Reply:
Thinking more about this, what happens if the resourceclaim that was allocated counted against quota from different ClusterQueue?

I think we have talked about this in the design review meeting that Kueue needs to support ResourceClaims that could be shared between workloads, but there is a grey area on how to count them against the right quota.

IMO, the correct thing to do here is: for all workloads that are using the allocated claim, count it against their quota only once.

cc @kannon92 @tardieu

Reply:
How is this handled with DRA's work for ResourceQuota?